packages/common/src/configuration/service-configuration.ts

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import * as convict from 'convict';
+import * as fs from 'fs';
+import { injectable } from 'inversify';
+import { isNil } from 'lodash';
+
+export interface CommonRuntimeConfig {
+    taskTimeoutInMinutes: number;
+    logInConsole: boolean;
+    maxQueueSize: number;
+}
+
+export interface ScanRunTimeConfig {
+    minLastReferenceSeenInDays: number;
+    pageRescanIntervalInDays: number;
+    failedPageRescanIntervalInHours: number;
+    maxScanRetryCount: number;
+}
+
+export interface RuntimeConfig {
+    commonConfig: CommonRuntimeConfig;
+    scanConfig: ScanRunTimeConfig;
+}
+
+@injectable()
+export class ServiceConfiguration {
+    public static readonly profilePath = './runtime-config/runtime-config.json';
+    private readonly fileSystem: typeof fs;
+    private loadConfigPromise: Promise<convict.Config<RuntimeConfig>>;
+    private readonly convictModule: typeof convict;
+
+    constructor(fileSystem: typeof fs = fs, convictModule: typeof convict) {
+        this.fileSystem = fileSystem;
+        this.convictModule = convictModule;
+    }
+
+    public async getConfigValue<K extends keyof RuntimeConfig | null | undefined = undefined>(
+        key?: K,
+    ): Promise<K extends null | undefined ? RuntimeConfig : RuntimeConfig[K]> {
+        const config = await this.getConvictConfig();
+
+        return config.get(key);
+    }
+
+    private async getConvictConfig(): Promise<convict.Config<RuntimeConfig>> {
+        if (isNil(this.loadConfigPromise)) {
+            this.loadConfigPromise = new Promise((resolve, reject) => {
```

Comment:
>Promise((resolve, reject) [](start = 41, length = 25)

Do we need to also reject on exception

Reply:
if the callback of promise throws then the promise will be rejected automatically. i have added a test to cover this now.